mysql-persistence/src/main/java/com/netflix/conductor/dao/mysql/MySQLBaseDAO.java

```diff
@@ -1,91 +1,202 @@
-		return String.format(queryTemplate, paramsPlaceholders);
-	}
+/**
+ * @author mustafa
+ */
+public abstract class MySQLBaseDAO {
+    private static final List<String> EXCLUDED_STACKTRACE_CLASS = ImmutableList.of(
+        "com.netflix.conductor.dao.mysql.MySQLBaseDAO", "java.lang.Thread");
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+    protected final ObjectMapper objectMapper;
+    protected final DataSource dataSource;
+
+    protected MySQLBaseDAO(ObjectMapper om, DataSource dataSource) {
+        this.objectMapper = om;
+        this.dataSource = dataSource;
+    }
+
+    protected final StackTraceElement getCaller() {
+        return Arrays.stream(Thread.currentThread().getStackTrace())
+                     .filter(ste -> !EXCLUDED_STACKTRACE_CLASS.contains(ste.getClassName()))
+                     .findFirst()
+                     .orElseThrow(() -> new NullPointerException("Cannot find Caller"));
+    }
+
+    protected String toJson(Object value) {
+        try {
+            return objectMapper.writeValueAsString(value);
+        } catch (JsonProcessingException ex) {
+            throw new ApplicationException(ApplicationException.Code.INTERNAL_ERROR, ex);
+        }
+    }
+
+    protected <T> T readValue(String json, Class<T> tClass) {
+        try {
+            return objectMapper.readValue(json, tClass);
+        } catch (IOException ex) {
+            throw new ApplicationException(ApplicationException.Code.INTERNAL_ERROR, ex);
+        }
+    }
+
+    protected <T> T readValue(String json, TypeReference<T> typeReference) {
+        try {
+            return objectMapper.readValue(json, typeReference);
+        } catch (IOException ex) {
+            throw new ApplicationException(ApplicationException.Code.INTERNAL_ERROR, ex);
+        }
+    }
+
+    /**
+     * Initialize a new transactional {@link Connection} from {@link #dataSource} and pass it to {@literal function}.
+     * <p/>
+     * Successful executions of {@literal function} will result in a commit and return of
+     * {@link TransactionalFunction#apply(Connection)}.
+     * <p/>
+     * If any {@link Throwable} thrown from {@code TransactionalFunction#apply(Connection)} will result in a rollback
+     * of the transaction
+     * and will be wrapped in an {@link ApplicationException} if it is not already one.
+     * <p/>
+     * Generally this is used to wrap multiple {@link #execute(Connection, String, ExecuteFunction)} or
+     * {@link #query(Connection, String, QueryFunction)} invocations that produce some expected return value.
+     *
+     * @param function The function to apply with a new transactional {@link Connection}
+     * @param <R>      The return type.
+     * @return The result of {@code TransactionalFunction#apply(Connection)}
+     * @throws ApplicationException If any errors occur.
+     */
+    protected <R> R getWithTransaction(TransactionalFunction<R> function) {
```

Comment:
I love what you did with the overall PR but this method I find it very hard to read

it's indeed difficult to manage all the flows of exceptions. The `Connection` object supports auto-closeable so it can be used in a try with resource block...

I tried to simplify the code a bit to improve readability, here's what I came up with:

```
protected <R> R getWithTransaction(TransactionalFunction<R> function) {
	Instant start = Instant.now();
	StackTraceElement caller = getCaller();
	logger.trace("{} : starting transaction", caller.getMethodName());

	try(Connection tx = dataSource.getConnection()) {
		boolean previousAutoCommitMode = tx.getAutoCommit();
		tx.setAutoCommit(false);
		try {
			R result = function.apply(tx);
			tx.commit();
			return result;
		} catch (SQLException ex) {
			tx.rollback();
			throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, ex);
		} finally {
			tx.setAutoCommit(previousAutoCommitMode);
		}
	} catch (SQLException ex) {
		throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, ex);
	} finally {
		logger.trace("{} : took {}ms", caller.getMethodName(), Duration.between(start, Instant.now()).toMillis());
	}
}
```

Reply:
This is much easier to read, the reason the checks for `logger.isTraceEnabled()` were being used is that this method gets called very frequently and I didn't want to burden the execution with parsing StackTraceElements if they weren't going to be used. 

I didn't spend a lot of time measuring what kind of execution cost `getCaller()` has, but I figured that if we are not logging `trace` level messages, then there is no reason to parse the `StackTraceElement`. This adds a lot of noise with if statements, but it gives us the best _likelihood_ of faster execution times.

I'm really not sure why I didn't use a try with resources on the connection, it was probably an oversight on my part. I also see that you added `previousAutoCommitMode` which is a fantastic idea.

Reply:
regarding the logger, since here we are using SLF4J with parameterized logging, the logger.isTraceEnabled() is redundant.

have a look at : https://www.slf4j.org/faq.html#logging_performance

Reply:
That's correct and we're using it, except for we are executing `getCaller()` no matter if the Logger is trace enabled or not. In my experience with this there's a trade off between a little harder to read and not executing debug/trace level calculations if we don't need to. In my other projects we usually have utility classes to handle this kind of stuff to reduce the boilerplate.

Again, I haven't measured the performance impact in any real detail but I assume that wrapping the `getCaller()` in a `isTraceEnabled()` block is more of a micro-optimization than a true performance benefit. I agree that it makes it harder to follow the code, so I think it's perfectly reasonable to follow your approach of removing the guard blocks and simplifying it.

Reply:
you are right @mashurex I overlooked that!

how about this: 

```
protected <R> R getWithTransaction(TransactionalFunction<R> function) {
        Instant start = Instant.now();
        LazyToString callingMethod = getLazyCallingMethod();
        logger.trace("{} : starting transaction", callingMethod);

        try(Connection tx = dataSource.getConnection()) {
            boolean previousAutoCommitMode = tx.getAutoCommit();
            tx.setAutoCommit(false);
            try {
                R result = function.apply(tx);
                tx.commit();
                return result;
            } catch (Throwable th) {
                tx.rollback();
                throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, th);
            } finally {
                tx.setAutoCommit(previousAutoCommitMode);
            }
        } catch (SQLException ex) {
            throw new ApplicationException(ApplicationException.Code.BACKEND_ERROR, ex);
        } finally {
            logger.trace("{} : took {}ms", callingMethod, Duration.between(start, Instant.now()).toMillis());
        }
    }

    protected final LazyToString getLazyCallingMethod() {
        return new LazyToString(() -> Arrays.stream(Thread.currentThread().getStackTrace())
            .filter(ste -> !EXCLUDED_STACKTRACE_CLASS.contains(ste.getClassName()))
            .findFirst()
            .map(StackTraceElement::getMethodName)
            .orElseThrow(() -> new NullPointerException("Cannot find Caller")));
    }
```

and the `LazyToString` class:

```
package com.netflix.conductor.dao.mysql;

import java.util.function.Supplier;

class LazyToString {
    private final Supplier<String> supplier;

    LazyToString(Supplier<String> supplier) {
        this.supplier = supplier;
    }

    @Override
    public String toString() {
        return supplier.get();
    }
}
```

Reply:
I really like this. I'll get it in today.